### ISSUE TITLE
Not working with HTTP proxy since 1.0.4

### ISSUE THREAD

**Issue by MichalMMac** [autopkg/autopkg#416]
I use AutoPkg on macOS which always uses HTTP(S) proxy for HTTP(S) connections.
Proxy is configured via environment variables. Since 1.0.4 I get following crash when doing `autopkg searxh xyz` or when running some recipes which connect to Github. This is probably related to change `GitHub API requests are now performed using curl. ` made in 1.0.4.

```
Traceback (most recent call last):
  File "/usr/local/bin/autopkg", line 2453, in <module>
    sys.exit(main(sys.argv))
  File "/usr/local/bin/autopkg", line 2447, in main
    exit(subcommands[verb]['function'](argv))
  File "/usr/local/bin/autopkg", line 912, in search_recipes
    results = do_gh_code_search(query, use_token=options.use_token)
  File "/usr/local/bin/autopkg", line 826, in do_gh_code_search
    accept="application/vnd.github.v3.text-match+json")
  File "/Library/AutoPkg/autopkglib/github/__init__.py", line 209, in call_api
    resp_data = json.loads(page_content)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/json/__init__.py", line 338, in loads
    return _default_decoder.decode(s)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/json/decoder.py", line 366, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/json/decoder.py", line 384, in raw_decode
    raise ValueError("No JSON object could be decoded")
ValueError: No JSON object could be decoded
``` 

macOS: 10.13.4
Clean install with comand-line tools installed.
Current workaround: Downgrade to 1.0.3

**Comment by Tugzrida** [comment]
Not working for me either. I did some investigating and the problem is in `autopkglib.github`. By switching to curl instead of urllib and manually parsing the headers, it appears to be getting confused by the headers returned when using a proxy. I added some verbose output to dump the entire raw data from the curl command this is the first bit of output:

    HTTP/1.1 200 Connection Established
    
    Proxy-Agent: Zscaler/6.0
    
    
    
    HTTP/1.1 200 OK
    
    Date: Fri, 04 May 2018 03:52:47 GMT
    ...

Running the same dump on an external system with no proxy yields exactly the same output(apart from datestamps changing, etc) EXCEPT it only has the standard `HTTP/1.1 200 OK` message(without the proxy's `HTTP/1.1 200 Connection Established` and headers) and DOES successfully decode the json.

I could fix this, but I'm not likely to have much time to right at the moment. Perhaps @hjuutilainen who wrote the curl code can help out? I'm happy to be a tester for anyone's proposed fixes. Will update if I do get a solution worked out.

**Comment by hjuutilainen** [comment]
Thanks for reporting this. I'll try to recreate the issue and see what can be done. The curl code is a simplified version of URLDownloader curl code and modified to parse stdout only. I can already see a couple of places where we could handle the proxy return code.

**Comment by gregneagle** [comment]
We should probably look at creating a base URLGetter processor that URLDownloader and others could subclass...

**Comment by jessepeterson** [comment]
We had the same issue over in reposado: wdas/reposado/pull/45 (which was an attempt that didn't work). The larger work happened in wdas/reposado/pull/47 which refactored the way the output processing loop worked for curl.

**Comment by Tugzrida** [comment]
Just did some further investigating and the output from curl looks like this (there's extra lines in the one above):

    HTTP/1.1 200 Connection Established
    Proxy-Agent: Zscaler/6.0
    
    HTTP/1.1 200 OK
    Date: Tue, 08 May 2018 04:27:33 GMT
    ...

The blank line catches on line 163 (`if info:`) and `donewithheaders` is set to true. `page_content` being passed to the json decoder then contains everything from `HTTP/1.1 200 OK` down, and the json decoder expects just a json object, so it fails. I could put a dirty fix in there but I'll leave it up to you guys to make it how you want :)

**Comment by jessepeterson** [comment]
@Tugzrida you could have just looked at the links I posted. :) There's code that addresses this curl output (in Python, too) that's been in use for over two years in another project. But you're right: project maintainers will have to decide what they want to do.

**Comment by MichalMMac** [comment]
I [refactored](https://github.com/MichalMMac/autopkg/tree/curlrefactor) the code fixing this bug in progress. Now I would love some [feedback](https://groups.google.com/forum/#!forum/autopkg-discuss).

**Comment by Tugzrida** [comment]
As far as I can tell, the refactored version works fine with the proxy, however the higher-ups have really messed around with how the proxy works here so I can't tell for sure. I may get some time during the Christmas break to have a more detailed look at it but at the moment I don't really have any time to spare :(

**Comment by kcrawford** [comment]
We are seeing this problem with our proxy and can confirm that the @MichalMMac branch resolves the issue.

**Comment by MichalMMac** [comment]
Pull request ready #446 Thank you for testing! :-)

**Comment by MichalMMac** [comment]
Changes were merged into [new_urlgetter](https://github.com/autopkg/autopkg/tree/new_urlgetter) branch. You can test if able.

**Comment by MichalMMac** [comment]
@Tugzrida @kcrawford Hi There. Could you test 1.4 (and/or 2.0b3) in your environment?
I believe this issue is now fixed and we can close it when verified.

**Comment by Tugzrida** [comment]
Sorry, I don’t have ready access to that environment anymore. I might be able to ask a colleague to try tomorrow.

**Comment by Tugzrida** [comment]
Turns out that the proxy setup in that environment has significantly changed so that it's not really reflective of the original bug here.

I did however just test v1.0.4 with a local test proxy to verify that I could reproduce the issue, then tried v1.4 and the issue is completely resolved as far as I can see.

Thanks for all your work @MichalMMac!

**Comment by MichalMMac** [comment]
I consider this fixed in 1.4 and 2.0 due to the refactored code in `URLGetter`.
